rfcs/sdk-js/RFC.md

```diff
@@ -0,0 +1,544 @@
+
+Runtime functionality will be:
+
+- `loadResource`: function that [loads a `Resource`](#loading-resources)
+- `changeLanguage`: function to change language
+- the [`lookup function`](#lookup-function)
+- a function to create [`alternate` links](#alternate-links)
+
+The `runtime` functions itself will not deal with reactivity. They just store `Resource`s in memory and the lookup function accesses them.
+
+### adapters
+
+On top of the core functionality we will have framework-specific implementations that will use the `runtime` functionality to deliver a tailored i18n experience.
+An adapter will provide a build plugin (vite, rollup, webpack, etc. via [`unplugin`](https://github.com/unjs/unplugin)). This build plugin will inject `runtime` functionality at specific parts of the codebase. Because most meta-frameworks have a convention-based setup, we can detect where to put things and how to best configure the framework to add i18n functionality. More information [here](#implementation-details-with-plugin)
+
+The adapter is also responsible to add reactivity to the `runtime` functionality where needed or if wanted by the developer (via a configuration option). An adapter will re-export the same functions as the `runtime` package does.
+
+Per default no reactivity gets used as it only makes sense for SPAs. If the language get's encoded into the url, then a simple redirect is better than to change the language client-side. Changing language is a rarely performed action and users will probably not expect that the page get's immediately updated. A page navigation is good enough.
+
+### loading resources
+
+Resources will be loaded from disk via the functions provided by `inlang.config.js`. Once loaded the `Resource`s will be kept in memory. A server then can transform (and in the future also optimize) those `Resource`s because it does not make sense to ship all `Resource`s to the client. This all happens during runtime, so we will probably also need to provide an `GET` endpoint so the client can load the optimized Resources from the server.
+
+In an ideal world we will always have all data needed. But in reality, things may be missing from a certain `Resource`. In that case, we transform the `Resource`s and include the reference if a `Id` is missing.
+
+### lookup function
+
+The lookup function will be the only thing that a developer needs to add to the source code. This function get's passed an `Id`, traverses the `AST` and returns a translated `Message`.
+
+The SDK will provide that function as the default export, so anyone can name it however he wants. We can also export some common aliases so users get useful auto-import capabilities from the IDE.
+
+```ts
+const lookupFunction = (key: string) => ""
+
+export default lookupFunction
+
+export const i18n = lookupFunction
+export const t = lookupFunction
+export const l = lookupFunction
+export const $ = lookupFunction // will not work well together with Svelte
+```
```

Comment:
I would vote for `i18n` as the name for the lookup function. Everyone (that has heard of internationalization) knows the term and should immediately know what this line does compared to `t` or `l`. It is also relatively fast to write and takes not that much space in the markup compared to `translate`. Inlang in the future (once the tooling is built) could also support inline strings instead of Id based translations. 

```ts
const name = 'Inlang'
i18n('welcome', name) // Id based
i18n`Welcome to this application, ${name}!` // inline
``` 

Further additions could include a way to add a description
```ts
i18n('welcome', name)
	.describe('Welcome message that shows up at the top of the page after completing the registration.')
```

Since `t` is often used in Id based libraries, it may not be that obvious that Inlangs solution offers more capabilities.

Reply:
I strongly vote for the lookup function.

1. Most teams will not use inlang to translate their product but enable non-technical users to make copy changes. Communicating "use the translate/i18n function" will confuse people "but, I don't want to translate anything". 
2. Lookup function is technically correct.

Reply:
so you would like to have `lookupFunction('welcome', name)` spread over the whole code base?

Reply:
Ah, no ofc not. I thought the conversation is about the debating whether the "lookup function" should be named "i18n". My mistake!

I had a shorthand `t` as lookup function in mind. `t` is shorter than `i18n` and we should try to avoid i18n to ease our communication of inlang [as in, can be used without translating apps].

Reply:
`t` for `translate` is also not a good fir then ^^
Maybe `i` or `in` for `inlang`?

Reply:
Hmm, why not `l` for lookup? I prefer the term `lookup`over `translate`, since I would expect a function named "translate" to do a translation of some string I have to provide... `i` would also be a good fit IMO, though a little undescriptive? Also, how about the classic `get`? ^^

Reply:
the small `l` looks like a big `I`. That could be confusing. I would not use `get`. In my opinion this looks a bit weird in your template ^^